llvm/include/llvm/ADT/STLExtras.h

```diff
@@ -1065,27 +1075,30 @@ class concat_iterator
   /// Returns null if the specified iterator is at the end. Otherwise,
   /// dereferences the iterator and returns the address of the resulting
   /// reference.
-  template <size_t Index> ValueT *getHelper() const {
+  template <size_t Index> handle_type getHelper() const {
     auto &Begin = std::get<Index>(Begins);
     auto &End = std::get<Index>(Ends);
     if (Begin == End)
-      return nullptr;
+      return {};
```

Comment:
Do we need this check at all? Should this be turned into an assertion instead or pushed to the callees?

Reply:
Not sure I understand the question/can picture what you're suggesting.

The helper function here is used to handle the variadic expansion of each kind of iterator/range being concatenated. I can't think of a particularly simpler way to do it?

Reply:
I think we could emit the extra optional by changing this helper a bit, and, e.g., having one helper for begin and one helper for end.

Then the loop that uses would look something like:
```c++
    // Loop over them, and return the first result we find.
    for (auto [GetBeginHelperFn, GetEndHelperFn] : GetHelperFns)
      auto &B = GetBeginHelperFn();
      auto &E = GetEndHelperFn();
      if (B != E)
        ...
```

But now I wrote it doesn't it doesn't seem any cleaner than the optional.